### DISCUSSION TITLE
Why do the validators need to return the validated values?

### DISCUSSION THREAD

**Discussion by tweakimp** [pydantic/pydantic#3012] [Question]
Why do the validators need to return the validated values?

This creates problems when I test validators that access the same field.
If one validator fails, the return value is missing and the field is not accessable any more, so the next validator fails as well.

Here is a minimal code example
```python
# klass.py
from pydantic import validator, root_validator, BaseModel

class Klass(BaseModel):
    string: str

    @root_validator
    def root_validator(cls, values):
        assert len(values["string"]) >= 3
        return values

    @validator("string")
    def validator(cls, string):
        assert "a" in string
        return string
```
Then the test file
```python
# test_klass.py
from klass import Klass

def test_1():
    Klass(string="test")
```

Now if I debug this test, it fails in two places!
The `validator` is evaluated first. Because there is `a in string` is `False`, the `assert` statement raises an `AssertionError`. This means the `return string` line is not executed.
Then the `root_validator` is evaluated, but fails with a `KeyError`, because `values` does not have the key `string`.

Why do the validators need to return to continue to validate the fields? Is it because I could change the values during validation? For example if I validate a time field provided as a string in `HH:MM` format, but then return it as a `datetime.time` object? But that would do more than just validate.

**Comment by lesnik512** [comment]


**Reply by tweakimp** [reply]
I know this is the correct behaviour, because it was built this way for a reason. But what was the reason?

Also, yes, asserts can be turned off with `-o`, but these assert statements are just in the tests. What would be your way of asserting test cases without assert?

**Comment by samuelcolvin** [comment]
> Why do the validators need to return the validated values?

Because they can (and very often do) modify the value.

See`return v.title()` in [the first example](https://pydantic-docs.helpmanual.io/usage/validators/).

**Reply by tweakimp** [reply]
Ok. I thought validators are only to validate, not to change the input.
